File: frontend/src/api/ipManagement.ts
/**
 * IP Management API functions
 */

import apiClient from './client';

export interface IPRangeAllocation {
  id: string;
  scenario_id: string;
  scenario_name: string;
  range_index: number;
  cidr_range: string;
  next_host_offset: number;
  created_at: string;
}

export interface IPRangeListResponse {
  items: IPRangeAllocation[];
  total: number;
  available_ranges: number[];
}

export interface NextIPResponse {
  ip_address: string;
  subnet_mask: string;
  gateway: string;
  cidr: string;
}

export interface ScenarioIPInfo {
  scenario_id: string;
  scenario_name: string;
  cidr_range: string;
  range_index: number;
  devices_with_ips: number;
  next_available_ip: string;
  gateway: string;
}

const IP_MANAGEMENT_PREFIX = '/api/v1/ip-management';

export const ipManagementApi = {
  /**
   * List all IP range allocations
   */
  async listAllocations(): Promise<IPRangeListResponse> {
    const response = await apiClient.get<IPRangeListResponse>(IP_MANAGEMENT_PREFIX);
    return response.data;
  },

  /**
   * Get IP info for a specific scenario
   */
  async getScenarioIPInfo(scenarioId: string): Promise<ScenarioIPInfo> {
    const response = await apiClient.get<ScenarioIPInfo>(
      `${IP_MANAGEMENT_PREFIX}/scenario/${scenarioId}`
    );
    return response.data;
  },

  /**
   * Get next available IP for a scenario (also increments the offset)
   */
  async getNextIP(scenarioId: string): Promise<NextIPResponse> {
    const response = await apiClient.get<NextIPResponse>(
      `${IP_MANAGEMENT_PREFIX}/scenario/${scenarioId}/next-ip`
    );
    return response.data;
  },
};

export default ipManagementApi;

File: backend/app/schemas/ip_management.py
"""IP Management schemas for API validation."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class IPRangeAllocationBase(BaseModel):
    """Base schema for IP range allocation."""

    range_index: int = Field(..., ge=1, le=254, description="Index in 10.{n}.0.0/16")
    cidr_range: str = Field(..., description="CIDR notation e.g. 10.1.0.0/16")
    next_host_offset: int = Field(default=10, ge=1, description="Next host offset")


class IPRangeAllocationResponse(IPRangeAllocationBase):
    """Response schema with scenario info."""

    id: UUID
    scenario_id: UUID
    scenario_name: str
    created_at: datetime

    class Config:
        from_attributes = True


class IPRangeListResponse(BaseModel):
    """List of all IP allocations."""

    items: list[IPRangeAllocationResponse]
    total: int
    available_ranges: list[int] = Field(
        description="Which range indices (1-254) are still free"
    )


class NextIPResponse(BaseModel):
    """Response for getting next available IP in a scenario's range."""

    ip_address: str = Field(..., description="Next available IP address")
    subnet_mask: str = Field(default="255.255.255.0", description="Subnet mask")
    gateway: str = Field(..., description="Gateway address for this subnet")
    cidr: str = Field(..., description="Scenario's CIDR range")


class ScenarioIPInfoResponse(BaseModel):
    """IP range info for a specific scenario."""

    scenario_id: UUID
    scenario_name: str
    cidr_range: str
    range_index: int
    devices_with_ips: int = Field(description="Count of devices that have IPs assigned")
    next_available_ip: str = Field(description="Next IP that will be assigned")
    gateway: str = Field(description="Default gateway for the range")
